### PR TITLE
Implement Sunnydale right to return DAH-790

### PR DESCRIPTION
**Pull request by akegan**
DAH-790

Still TODO:

- [x] Make sure address verfication still works for AG, but is not turned on for Sunnydale
- [x] Add SFHA lease pref proof option
- [x] Add sfha_letter_instructions content once we decide on it
- [x] Get validation up and running
- [x] Update/fix tests
- [x] Update doc checklist
- [x] Make sure listing details page links to document checklist
- [x] Push up changes for translation
- [x] Add SF City ID proof option, Update label for SFHA letter to use "residency"
- [x] Rename url to make it more generic
- [x]  Add more tests

### CODE REVIEW

**Comment by akegan** [comment]
@james-wills-sf this is ready to start reviewing. Last pieces are:
- I need to do a phrase push/pull and rename some keys
- I want to add some new tests for some of the functions I added.

If anything doesn't make sense let me know! I can definitely add more comments. Lots of changes had to be made to get the 🍝  to work.

File: app/assets/javascripts/short-form/ShortFormHelperService.js.coffee [hunk 1]
```diff
@@ -305,8 +317,11 @@ ShortFormHelperService = ($translate, $filter, $sce, $state) ->
         Service.preference_proof_options_live
       when 'rentBurden'
         Service.preference_proof_options_rent_burden
+      # FIXME: update to make this automatic.
```
Comment (akegan) [inline app/assets/javascripts/short-form/ShortFormHelperService.js.coffee:320]: 🤔  will have to come back to this. I think it's probably OK now

File: app/assets/javascripts/short-form/templates/e6a-right-to-return-preference.html.slim [hunk 2]
```diff
@@ -0,0 +1,39 @@
+.app-card.form-card
```
Comment (akegan) [inline app/assets/javascripts/short-form/templates/e6a-right-to-return-preference.html.slim:1]: This was app/assets/javascripts/short-form/templates/e6a-alice-griffith-preference.html.slim before but I renamed it and the diff was big enough that github didn't pick it up as a rename

File: app/assets/json/translations/locale-en.json [hunk 3]
```diff
@@ -293,6 +293,8 @@
             "rent_burden_doc_2": "You'll also need to show last month's rent with one of the following:",
             "rent_burden_title": "Rent Burdened Preference",
             "resident_desc": "You'll need one of the following documents showing the address where you live in San Francisco:",
+            "right_to_return_title": "HOPE SF Right to Return Preference",
+            "right_to_return_desc": "You'll need one of the following documents showing your address at Sunnydale Housing Development.",
```
Comment (akegan) [inline app/assets/json/translations/locale-en.json:297]:
```suggestion
```
These aren't being used anymore. Would come out in the grunt translations run I'm about to do anyways

File: spec/javascripts/services/short_form_data_service_spec.coffee [hunk 4]
```diff
@@ -8,37 +8,40 @@
...
+      $provide.value 'SharedService', jasmine.createSpy()
+      $provide.value 'ListingLotteryService', jasmine.createSpy()
+      $provide.value 'ExternalTranslateService', jasmine.createSpy()
+      $provide.value 'ListingEligibilityService', jasmine.createSpy()
+      $provide.value 'IncomeCalculatorService', jasmine.createSpy()
       return
     )
 
-    beforeEach inject((_ShortFormDataService_) ->
+    beforeEach inject((_ShortFormDataService_, _ListingDataService_, _ListingConstantsService_, _$state_, _$httpBackend_, _$localStorage_) ->
```
Comment (akegan) [inline spec/javascripts/services/short_form_data_service_spec.coffee:36]: Marçin finally helped me figure out how to pull real modules into tests. The syntax is terrible and you have to add all the dependencies above, but it's better than trying to constantly mock everything

File: spec/javascripts/fixtures/json/listings-api-listing-preferences.json [hunk 5]
```diff
@@ -1 +1 @@
...
...:5,"requiresProof":false,"readMoreUrl":"http://sfmohcd.org/housing-preference-programs","preferenceProofRequirementDescription":null,"preferenceName":"Live or Work in San Francisco Preference","pdfUrl":null,"order":5,"NRHPDistrict":null,"name":"LP-09778","listingPreferenceID":"a0l0P00001Lx8XeQAJ","listingId":"a0W0P00000F8YG4UAN","description":"For households in which at least one member lives or works in San Francisco.  Requires submission of proof.  Please note in order to claim Work Preference, the applicant currently work in San Francisco at least 75% of their working hours.","appTotal":null},{"unitsAvailable":null,"requiresProof":false,"readMoreUrl":null,"preferenceProofRequirementDescription":null,"preferenceName":"Works in Public Ed","pdfUrl":null,"order":null,"NRHPDistrict":null,"name":"LP-32885","listingPreferenceID":"a0l0P00001PsqDoQAJ","listingId":"a0W0P00000F8YG4UAN","description":"For households in which at least one member was a resident of the Alice Griffith housing development. This includes baseline and current residents that lived in the targeted redevelopment site on or after the time of application for Choice Neighborhoods of October 26, 2010.","appTotal":null}]}
```
Comment (akegan) [inline spec/javascripts/fixtures/json/listings-api-listing-preferences.json:1]: Changed Alice Griffith named preference to "Works in public ed" here to get it to work with a test that was expecting a custom preference.

**Review by akegan — COMMENTED** [review]
Couple of comments

**Review by james-wills-sf — COMMENTED** [review]
@akegan have a few questions and only one blocking bug

### Question: I see RTR as the 5th preference option on the listing details page, is that right?
Also possible it's just the way the test listing was set up and it will be different in prod

### [maybe major] Bug: RTR preference proof not uploaded
I submitted two applications with two separate accounts with L/W proof and sunnydale proofs attached, but after submit you can only see the L/W proof.

See this application has only one attached proof: https://sfhousing--full.lightning.force.com/lightning/r/Application__c/a0o6s0000002SNBAA2/view

## [minor, maybe related] Bug: Apt text gets shoved into main address after save
Could this be related to repeated apt numbers DAH-817
- with an account, fill out application with RTR pref
- Add an address
- Add an apartment number "Apt 1" in the apartment number field
- click "save and finish later"
- Resume editing the application
- Note that now "Apt 1" is part of the main address field
- 2021-05-25 at 11 48 43 AM


### [minor, unrelated to this change] Bug: view submitted application in new tab
This is likely unrelated to your changes, just maybe a behavior we can fix in the rewrite.
- without an account, fill out an application
- submit the application
- on the application submitted page, right click "View submitted application and print a copy" and open in a new tab
- 2021-05-25 at 11 35 40 AM

- Note that the new tab redirects to the homepage instead of showing the submitted app

**Comment by akegan** [comment]
An update. I've done my Phrase updates and addressed comments. Here are some answers to questions too! We can wait til Larry has done DAH-821 to re-test

> Question: I see RTR as the 5th preference option on the listing details page, is that right?

Yep, for this listing, it's showing up 5th because I was too lazy to renumber all the preferences. When Mel creates her listing she can give it the right rank, it doesn't affect the application flow

> [maybe major] Bug: RTR preference proof not uploaded

Good catch! Created DAH-821 for Larry to fix

> [minor, maybe related] Bug: Apt text gets shoved into main address after save

Salesforce doesn't store them as separate fields so this is expected. I believe the 2nd address line is to help with Easypost address validation which we're not doing that for this preference. It shouldn't impact what's saved. 

> [minor, unrelated to this change] Bug: view submitted application in new tab

I made DAH-822 for the issue of opening the application in a new tab to print

File: app/assets/javascripts/short-form/ShortFormApplicationController.js.coffee [hunk 6]
```diff
@@ -468,13 +468,41 @@ ShortFormApplicationController = (
...
+
+  $scope.rtrTranslationKey =
+    switch $scope.getRTRPreference()
+      when 'rightToReturnSunnydale'
+        'rtr_sunnydale'
+      when 'aliceGriffith'
+        'alice_griffith'
+
+
+  $scope.preferenceDesc = $translate.instant("preferences.#{$scope.rtrTranslationKey}.desc")
```
[the remark this answers is not in the document]
Comment (akegan) [reply app/assets/javascripts/short-form/ShortFormApplicationController.js.coffee]: fixed

File: app/assets/javascripts/listings/ListingPreferenceService.js.coffee [hunk 7]
```diff
@@ -42,6 +42,21 @@ ListingPreferenceService = ($http, ListingConstantsService, ListingIdentityServi
...
+  Service.isRTRPreference = (preferenceName) ->
+    # Determine if preference is one of the right to return preferences
+    return _.includes(ListingConstantsService.rightToReturnPreferences, preferenceName)
+
+  Service.hasRTRPreference = (listing) ->
+    # Determine if the listing has a right to return preference
+    hasRTRMap = _.map(ListingConstantsService.rightToReturnPreferences, (pref) -> Service.hasPreference(pref, listing))
+    _.some(hasRTRMap)
+
+  Service.getRTRPreference = (listing) ->
```
[the remark this answers is not in the document]
Comment (akegan) [reply app/assets/javascripts/listings/ListingPreferenceService.js.coffee]: done!

**Review by james-wills-sf — APPROVED** [review]
@akegan lgtm!

**Pull request merged**